go/runtime/txpool/txpool.go

```diff
@@ -581,31 +614,34 @@ func (t *txPool) checkTxBatch(ctx context.Context, rr host.RichRuntime) {
 		// Notify submitter of success.
 		notifySubmitter(batchIndices[i])
 
-		// Publish local transactions immediately.
-		publishTime := time.Now()
-		if tx.flags.isLocal() {
-			if err := t.txPublisher.PublishTx(ctx, tx.tx); err != nil {
-				t.logger.Warn("failed to publish local transaction",
-					"err", err,
-					"tx_hash", tx.hash,
-				)
-
-				// Since publication failed, make sure we retry early.
-				t.republishCh.In() <- struct{}{}
-				publishTime = time.Time{}
+		if !pct.flags.isRecheck() {
+			// Mark new transactions as never having been published. The republish worker will
+			// publish these immediately.
+			publishTime := time.Time{}
+			if pct.dstQueue == t.mainQueue {
+				// This being a tx we got from outside, it's usually something that another node
+				// has just broadcast. Treat it as if it were published just now so that we don't
+				// immediately publish again from our node.
+				publishTime = time.Now()
 			}
+			// Put cannot fail as seenCache's LRU capacity is not in bytes and the only case where it
+			// can error is if the capacity is in bytes and the value size is over capacity.
+			_ = t.seenCache.Put(pct.Hash, publishTime)
 		}
+	}
 
-		// Put cannot fail as seenCache's LRU capacity is not in bytes and the only case where it
-		// can error is if the capacity is in bytes and the value size is over capacity.
-		_ = t.seenCache.Put(tx.hash, publishTime)
+	if numNewTxs != 0 {
+		// Kick off publishing for any new txs.
+		t.republishCh.In() <- struct{}{}
 	}
 
 	// Notify subscribers that we have received new transactions.
-	t.checkTxNotifier.Broadcast(newTxs)
+	// todo: this now broadcasts rechecked transactions too
```

Comment:
oh yeah I'll change this back. our only subscriber on this works to fill in the proposed transactions, which won't go wrong, but it would be a waste to have it do extra passes over txs that it already has